backend/backend/endpoints/websocket/WebSocketHandler.cs

```diff
@@ -0,0 +1,281 @@
+using System.Text.Json.Serialization;
+using System.Collections.Concurrent;
+using Microsoft.AspNetCore.Authorization.Infrastructure;
+using System.Diagnostics;
+
+namespace backend.endpoints.websocket;
+
+using MessageType = string;
+
+public struct TypedMessage<T>
+{
+    [JsonPropertyName("type")]
+    public MessageType Type { get; set; }
+
+    [JsonPropertyName("msg")]
+    public T Msg { get; set; }
+}
+
+public delegate Task MessageHandlerDelegate(string clientToken, JsonElement messageData);
+public delegate Task TypedMessageHandlerDelegate<T>(string clientToken, T messageData);
+
+public static class WebSocketHandler
+{
+    private static readonly ConcurrentDictionary<string, WebSocket> ActiveConnections = new();
+    private static readonly Random Random = new();
+    private static readonly ConcurrentDictionary<MessageType, List<MessageHandlerDelegate>> MessageHandlers = new();
+
+    private static string GenerateClientToken()
+    {
+        const string chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
+        return new string([.. Enumerable.Repeat(chars, 5).Select(s => s[Random.Next(s.Length)])]);
+    }
+
+    /// <summary>
+    /// Generates a unique client ID and adds the WebSocket connection to the active connections list. Returns the client ID.
+    /// </summary>
+    private static string AddConnection(WebSocket webSocket)
+    {
+        string clientToken;
+        do
+        {
+            clientToken = GenerateClientToken();
+        } while (!ActiveConnections.TryAdd(clientToken, webSocket));
+
+        return clientToken;
+    }
+
+    /// <summary>
+    /// Removes a WebSocket connection from the active connections list. The client token must be valid and the client connected. 
+    /// </summary>
+    private static void RemoveConnection(string clientToken)
+    {
+        var result = ActiveConnections.TryRemove(clientToken, out _);
+
+        Debug.Assert(result, $"Failed to remove client with ID: {clientToken}");
+    }
+
+    /// <summary>
+    /// Sends a typed message to a specific client. The client ID must be valid and connected. Returns true if the message was sent successfully, false otherwise.
+    /// </summary>
+    public static async Task<bool> SendMessage<T>(string clientToken, TypedMessage<T> message)
+    {
+        var result = ActiveConnections.TryGetValue(clientToken, out var webSocket);
+
+        Debug.Assert(result, $"Failed to find client with ID: {clientToken}");
+        Debug.Assert(webSocket != null);
+
+        try
+        {
+            var messageJson = JsonSerializer.Serialize(message);
+            var messageBytes = Encoding.UTF8.GetBytes(messageJson);
+            await webSocket.SendAsync(
+                new ArraySegment<byte>(messageBytes),
+                WebSocketMessageType.Text,
+                true,
+                CancellationToken.None);
+
+            return true;
+        }
+        catch (Exception)
+        {
+            return false;
+        }
```

Comment:
fragwuerdige fehlerbehandlung
```suggestion
        catch (WebSocketException ex)
        {
            Console.Error.WriteLine($"WebSocket error with {clientToken}: {ex.Message}");
            webSocket.Abort();
            ActiveConnections.TryRemove(clientToken, out _);
        }
        catch (ObjectDisposedException ex)
        {
            Console.Error.WriteLine($"Attempted to use a disposed WebSocket for {clientToken}: {ex.Message}");
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Unexpected error sending message to {clientToken}: {ex}");
        }
        return false;
```

Reply:
fuers entfernen vllt eher unser wrapper statt ein direkter tryremove. abort sollte aber korrekt sein fuer einen fehlerhaften socket.

Reply:
Ich sehe nicht, wo hier `WebSocketException` oder `ObjectDisposedException` ausgelöst werden? 

`TryGetValue(...)` löst ggf. eine 'ArgumentNullException', da `clientToken` aber niemals `null` ist, wird das nicht eintreffen. 

`Serialize(...)` löst ggf. eine `NotSupportedException` aus. Würde TypedMessage ein Konstrukt enthalten, das nicht Json serializable ist, wird diese ausgelöst. 

`GetBytes(...)` löst `ArgumentNullException` oder `EncoderFallbackException` aus. `ArgumentNullException` geht nicht, weil das Argument nie `null` ist. `EncoderFallbackException` wird nicht ausgelöst, weil `System.Text.Encoding.EncoderFallback` standardmäßig `System.Text.EncoderReplacementFallback` und nicht `System.Text.EncoderExceptionFallback` ist. Siehe dazu auch [hier](https://learn.microsoft.com/en-us/dotnet/api/system.text.encoderfallback?view=net-9.0).

[`SendAsync(...)`](https://learn.microsoft.com/en-us/dotnet/api/system.net.websockets.websocket.sendasync?view=net-9.0#exceptions) kann nur `OperationCanceledException` auslösen. Da aber `CancellationToken.None` als Parameter übergeben wird, tritt dieser Fall nie ein.

Schließlich kann der Konstruktor von `ArraySegment` eine `ArgumentNullException` auslösen. `messageBytes` ist aber niemals `null`.

Effektiv tritt der Exception Fall also nur auf, wenn `message` ein Konstrukt enthält, das nicht JSON serializable ist. Oder, wenn der Standard-Wert von `EncoderFallback` geändert wird. In beiden Fällen wird `false` als Rückgabe geliefert, was, dem Schnittstellendefinition entsprechend, einen Nicht-erfolgreichen Senden entspricht. Besonders denke ich, dass die WebSocket Verbindung nicht automatisch geschlossen werden sollte, wenn ein solcher Fehler auftritt, da so unterwartet die Verbindung invalid wird.